PyInstaller/archive/readers.py

```diff
@@ -143,15 +143,23 @@ def checkmagic(self):
         else:
             self.lib.seek(0, 2)
         filelen = self.lib.tell()
-        if self.length:
-            self.lib.seek(self.start + self.length - self._cookie_size, 0)
-        else:
-            self.lib.seek(-self._cookie_size, 2)
-        (magic, totallen, tocpos, toclen, pyvers, pylib_name) = struct.unpack(
+        found = False
+        for i in range(4096):
+            self.lib.seek(filelen-self._cookie_size, 0)
+            (magic, totallen, tocpos, toclen, pyvers, pylib_name) = struct.unpack(
```

Comment:
Hmm, thinking about this: Maybe we should take the chance and change the C-code, too?